<a href="https://colab.research.google.com/github/Seif-Eltaweel/Seif-Data-Science-Portfolio/blob/main/Movie_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The TMDb database

In this supplementary exam, you will be exploring [The Movie Database](https://www.themoviedb.org/) – an online movie and TV show database that houses some of the most popular movies and TV shows at your fingertips. The TMDb database supports 39 official languages used in over 180 countries daily and dates back all the way to 2008.


<img src="https://github.com/Explore-AI/Pictures/blob/master/sql_tmdb.jpg?raw=true" width=80%/>


Below is an Entity Relationship Diagram (ERD) of the TMDb database:

<img src="https://github.com/Explore-AI/Pictures/blob/master/TMDB_ER_diagram.png?raw=true" width=70%/>

As can be seen from the ERD, the TMDb database consists of `12 tables` containing information about movies, cast, genre, and so much more.  

Let's get started!

###Loading the data



In [1]:
'/content/TMDB-a-4006.db'

'/content/TMDB-a-4006.db'

In [2]:
%reload_ext sql

In [3]:
%sql sqlite:///TMDB-a-4006.db

### Question 1
Who won the Oscar for “Actor in a Leading Role” in 2015?

(Hint: The winner is indicated as '1.0'.)

In [5]:
%sql PRAGMA integrity_check;

 * sqlite:///TMDB-a-4006.db
Done.


[('ok',)]

In [6]:
%%sql
   SELECT * FROM oscars
   WHERE year = 2015
   AND award LIKE 'Actor in a Leading Role'
   AND Winner = '1.0'
   LIMIT 15;

 * sqlite:///TMDB-a-4006.db
Done.


[('2015', 'Actor in a Leading Role', '1.0', 'Leonardo DiCaprio', 'The Revenant')]

So the asnswer is **Leonardo DiCaprio**

###Question 2

What query will produce the ten oldest movies in the database?

In [10]:
%%sql
SELECT * FROM movies WHERE release_date IS NOT NULL ORDER BY release_date ASC LIMIT 10 ;

 * sqlite:///TMDB-a-4006.db
Done.


[(3059, 'Intolerance', '1916-09-04 00:00:00.000000', 385907, None, 'en', 'Intolerance', 'The story of a poor young woman, separated by prejudice from her husband and baby, is interwoven with tales of intolerance from throughout history.', 3.232447, 8394751.0, 197.0, 'Released', 'The Cruel Hand of Intolerance', 7.4, 60),
 (3060, 'The Big Parade', '1925-11-05 00:00:00.000000', 245000, None, 'en', 'The Big Parade', "The story of an idle rich boy who joins the US Army's Rainbow Division and is sent to France to fight in World War I, becomes friends with two working class men, experiences the horrors of trench warfare, and finds love with a French girl.", 0.785744, 22000000.0, 151.0, 'Released', None, 7.0, 21),
 (19, 'Metropolis', '1927-01-10 00:00:00.000000', 92620000, None, 'de', 'Metropolis', "In a futuristic city sharply divided between the working class and the city planners, the son of the city's mastermind falls in love with a working class prophet who predicts the coming of a savior to mediate their differences.", 32.351527, 650422.0, 153.0, 'Released', 'There can be no understanding between the hands and the brain unless the heart acts as mediator.', 8.0, 657),
 (905, "Pandora's Box", '1929-01-30 00:00:00.000000', 0, None, 'de', 'Die Bnchse der Pandora', 'The rise and inevitable fall of an amoral but naive young woman whose insouciant eroticism inspires lust and violence in those around her.', 1.824184, 0.0, 109.0, 'Released', None, 7.6, 45),
 (65203, 'The Broadway Melody', '1929-02-08 00:00:00.000000', 379000, None, 'en', 'The Broadway Melody', "Harriet and Queenie Mahoney, a vaudeville act, come to Broadway, where their friend Eddie Kerns needs them for his number in one of Francis Zanfield' ... (195 characters truncated) ... eenie recognizes, that she is for Jock nothing more than a toy, and it also takes a while till Harriet recognizes, that Eddie is in love with Queenie", 0.968865, 4358000.0, 100.0, 'Released', "The pulsating drama of Broadway's bared heart speaks and sings with a voice to stir your soul!", 5.0, 19),
 (22301, "Hell's Angels", '1930-11-15 00:00:00.000000', 3950000, None, 'en', "Hell's Angels", "Two brothers attending Oxford enlist with the Royal Flying Corps when World War I breaks out. Roy and Monte Rutledge have very different personalitie ... (576 characters truncated) ... t a strategic German munitions facility is a booming success, but with a squadron of fighters bearing down on them afterwards, escape seems unlikely.", 8.484123, 8000000.0, 127.0, 'Released', "Howard Hughes' Thrilling Multi-Million Dollar Air Spectacle", 6.1, 19),
 (22649, 'A Farewell to Arms', '1932-12-08 00:00:00.000000', 4, None, 'en', 'A Farewell to Arms', "British nurse Catherine Barkley (Helen Hayes) and American Lieutenant Frederic Henry (Gary Cooper) fall in love during the First World War in Italy.  ... (105 characters truncated) ... the war rages on. Academy Awards winner for Best Cinematography and for Best Sound, Recording. Nominated for Best Picture and for Best Art Direction.", 1.199451, 25.0, 89.0, 'Released', 'Every woman who has loved will understand', 6.2, 28),
 (3062, '42nd Street', '1933-02-02 00:00:00.000000', 439000, None, 'en', '42nd Street', 'A producer puts on what may be his last Broadway show, and at the last moment a chorus girl has to replace the star.', 1.933366, 2281000.0, 89.0, 'Released', None, 6.1, 37),
 (43595, 'She Done Him Wrong', '1933-02-09 00:00:00.000000', 200000, None, 'en', 'She Done Him Wrong', '"New York singer and nightclub owner Lady Lou has more men friends than you can imagine. Unfortunately one of them is a vicious criminal who\'s escap ... (52 characters truncated) ... ing she hasn\'t exactly been faithful in his absence. Help is at hand in the form of young Captain Cummings a local temperance league leader though."', 0.622752, 2200000.0, 66.0, 'Released', "Mae West gives a 'Hot Time' to the nation!", 5.1, 27),
 (3078, 'It Happened One Night', '1934-02-22 00:00:00.000000', 325000,

The answer is : SELECT * FROM movies WHERE release_date IS NOT NULL ORDER BY release_date ASC LIMIT 10 ;

###Question 3
How many unique awards are there in the Oscars table?

In [7]:
%%sql
SELECT COUNT(DISTINCT award) AS unique_awards FROM oscars;


 * sqlite:///TMDB-a-4006.db
Done.


[(114,)]

The answer is : **114**

### Question 4

How many movies are there that contain the word “Spider” within their title?

In [15]:
%%sql
    Select count(title) as spider_title
    from movies
    where title like '%spider%' ;

 * sqlite:///TMDB-a-4006.db
Done.


[(9,)]

### Question 5

How many movies are there that are both in the "Thriller" genre and contain the word “love” anywhere in the keywords?



The answer is : **9**

In [16]:
%%sql
      SELECT count(movies.movie_id) as love_thrilled_movies
      from movies
        JOIN genremap
        ON genremap.movie_id = movies.movie_id
        join keywordmap
        on keywordmap.movie_id = movies.movie_id

where genremap.genre_id =53
and keyword_id = 9673
;


 * sqlite:///TMDB-a-4006.db
Done.


[(14,)]

The answer is : **14**

### Question 6

How many movies are there that were released between 1 August 2006 ('2006-08-01') and 1 October 2009 ('2009-10-01') that have a popularity score of more than 40 and a budget of less than 50 000 000?

In [17]:
%%sql
     SELECT COUNT(*)
FROM movies
WHERE release_date BETWEEN '2006-08-01' AND '2009-10-01'
AND popularity > 40
AND budget < 50000000
;


 * sqlite:///TMDB-a-4006.db
Done.


[(29,)]

The answer is : **29**

### Question 7

How many unique characters has "Vin Diesel" played so far in the database?


In [18]:
%%sql
     select count(DISTINCT characters)
    from Casts
    where actor_id = 12835
 ;

 * sqlite:///TMDB-a-4006.db
Done.


[(16,)]

The answer is : **16**

### Question 8

What are the genres of the movie “The Royal Tenenbaums”?

In [19]:
%%sql
select DISTINCT * from genres where genre_id = 18 or genre_id = 35 ;

 * sqlite:///TMDB-a-4006.db
Done.


[(18, 'Drama'), (35, 'Comedy')]

The abswer is : **Drama, Comedy**

### Question 9

What are the three production companies that have the highest movie popularity score on average, as recorded within the database?

In [20]:
%%sql
    select  movies.movie_id,
            avg(movies.popularity) as avg_popularity,
            productioncompanymap.production_company_id,
            productioncompanies.production_company_name
    from  movies
    join  productioncompanymap
    on    movies.movie_id = productioncompanymap.movie_id
    join  productioncompanies
    on    productioncompanies.production_company_id =  productioncompanymap.production_company_id
    group by production_company_name
    order by avg_popularity  desc
    limit 3 ;



 * sqlite:///TMDB-a-4006.db
Done.


[(293660, 514.569956, 11307, "The Donners' Company"),
 (118340, 481.098624, 54850, 'Bulletproof Cupid'),
 (246655, 326.92099900000005, 78091, 'Kinberg Genre')]

The answer is : **The Donners' Company > Bulletproof Cupid > Kinberg Genre**

### Question 10

How many female actors (i.e. gender = 1) have a name that starts with the letter "N"?


In [24]:
%%sql
     select count(actor_name) as female_with_N
     from actors
        where gender = 1
        and actor_name like 'N%';

 * sqlite:///TMDB-a-4006.db
(sqlite3.OperationalError) no such table: actors
[SQL: select count(actor_name) as female_with_N
     from actors
        where gender = 1 
        and actor_name like 'N%';]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


The answer is : **355**

### Question 11

Which genre has, on average, the lowest movie popularity score?


In [8]:
%%sql
     select avg(movies.popularity) as avg_popularity,
            movies.movie_id,
            genremap.genre_id,
            genres.genre_name
    from movies
    join genremap
    on   movies.movie_id = genremap.movie_id
    join genres
    on   genremap.genre_id = genres.genre_id
    group by genres.genre_name
    order by avg_popularity asc ;

 * sqlite:///TMDB-a-4006.db
Done.


[(0.686786794117647, 7509, 10769, 'Foreign'),
 (3.945723809090908, 291, 99, 'Documentary'),
 (6.389414749999999, 10947, 10770, 'TV Movie'),
 (13.101512475675676, 16, 10402, 'Music'),
 (15.962426457494416, 13, 10749, 'Romance'),
 (17.444838720812175, 79, 36, 'History'),
 (17.764853371789304, 13, 18, 'Drama'),
 (18.221000988966285, 5, 35, 'Comedy'),
 (18.23627854878048, 33, 37, 'Western'),
 (18.295915104046244, 170, 27, 'Horror'),
 (22.85327434051726, 5, 80, 'Crime'),
 (23.777288569444423, 25, 10752, 'War'),
 (24.460076525117703, 18, 53, 'Thriller'),
 (24.58682736494254, 62, 9648, 'Mystery'),
 (27.832848912280713, 12, 10751, 'Family'),
 (30.94038215337952, 11, 28, 'Action'),
 (36.38704340094338, 18, 14, 'Fantasy'),
 (36.45180587102805, 11, 878, 'Science Fiction'),
 (38.81343931196578, 12, 16, 'Animation'),
 (39.26804199113923, 11, 12, 'Adventure')]

The answer is : **Foreign**

### Question 12

Which award category has the highest number of actor nominations (actors can be male or female)? (Hint: `Oscars.name` contains both actors' names and film names.)


In [9]:
%%sql
SELECT oscars.award, COUNT(oscars.name) AS nominations
FROM oscars
JOIN actors ON oscars.name = actors.actor_name
WHERE oscars.winner IS NOT NULL
GROUP BY oscars.award
ORDER BY nominations DESC
LIMIT 10;


 * sqlite:///TMDB-a-4006.db
Done.


[('Actor in a Supporting Role', 74),
 ('Actress in a Supporting Role', 71),
 ('Actor', 41),
 ('Actor in a Leading Role', 39),
 ('Actress in a Leading Role', 38),
 ('Honorary Award', 35),
 ('Actress', 32),
 ('Jean Hersholt Humanitarian Award', 15),
 ('Special Award', 14),
 ('Irving G. Thalberg Memorial Award', 12)]

The answer is : **Actor in a Supporting Role**